### PR TITLE
Introduce fault controller

### PR DESCRIPTION
**Pull request by livanov93**

### CODE REVIEW

File: kortex2_bringup/launch/kortex_control.launch.py [hunk 1]
```diff
@@ -167,6 +167,13 @@ def generate_launch_description():
             description="Robot hand controller to start.",
         )
     )
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            "fault_controller",
+            default_value="fault_controller",
+            description="Robot hand controller to start.",
```
Comment (destogl) [inline kortex2_bringup/launch/kortex_control.launch.py]:
```suggestion
            description="Name of the 'fault controller'.",
```

File: kortex2_controllers/CMakeLists.txt [hunk 2]
```diff
@@ -10,16 +10,20 @@ find_package(ament_cmake REQUIRED)
 find_package(ament_cmake_ros REQUIRED)
 find_package(controller_interface REQUIRED)
 find_package(geometry_msgs REQUIRED)
+find_package(std_msgs REQUIRED)
+find_package(std_srvs REQUIRED)
```
Comment (destogl) [inline kortex2_controllers/CMakeLists.txt]: These packages are deprecated. We should probably use then "example_msgs" package instead.

File: kortex2_controllers/src/fault_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  return config;
+}
+
+controller_interface::InterfaceConfiguration FaultController::state_interface_configuration() const
+{
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  config.names.emplace_back("reset_fault/in_fault");
```
Comment (destogl) [inline kortex2_controllers/src/fault_controller.cpp]: "in_fault" is a bit confusing name. Can you make it more descriptive? Does this mean "internal_fault"?

File: kortex2_controllers/include/kortex2_controllers/fault_controller.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  CallbackReturn on_deactivate(const rclcpp_lifecycle::State & previous_state) override;
+
+  KORTEX2_CONTROLLERS_PUBLIC
+  controller_interface::return_type update(
+    const rclcpp::Time & time, const rclcpp::Duration & period) override;
+
+private:
+  bool resetFault(const CmdType::Request::SharedPtr req, CmdType::Response::SharedPtr resp);
+
+  rclcpp::Publisher<FbkType>::SharedPtr fault_pub_;
```
Comment (destogl) [inline kortex2_controllers/include/kortex2_controllers/fault_controller.hpp:83]: You should use Realtime publisher. See example in broadcasters.

File: kortex2_controllers/src/fault_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  return CallbackReturn::SUCCESS;
+}
+
+bool FaultController::resetFault(
+  const CmdType::Request::SharedPtr /*req*/, CmdType::Response::SharedPtr resp)
+{
+  command_interfaces_[CommandInterfaces::RESET_FAULT_ASYNC_SUCCESS].set_value(ASYNC_WAITING);
+  command_interfaces_[CommandInterfaces::RESET_FAULT_CMD].set_value(ISSUE_CMD);
+
+  RCLCPP_INFO(node_->get_logger(), "Trying to reset faults on kinova controller.");
```
Comment (destogl) [inline kortex2_controllers/src/fault_controller.cpp:117]: If you find this output useful, then also an output that the fault is resetted could be useful.

**Review by destogl — CHANGES_REQUESTED** [review]
The most important change would be to use Realtime Publisher in the controller. Other changes are rather cosmetic.

File: kortex2_controllers/include/kortex2_controllers/fault_controller.hpp (hunk 4 above)
Comment (livanov93) [reply kortex2_controllers/include/kortex2_controllers/fault_controller.hpp:83]: 3bfe4b5

File: kortex2_controllers/src/fault_controller.cpp (hunk 3 above)
Comment (livanov93) [reply kortex2_controllers/src/fault_controller.cpp]: ab09b14

File: kortex2_controllers/CMakeLists.txt (hunk 2 above)
Comment (livanov93) [reply kortex2_controllers/CMakeLists.txt]: ab09b14

File: kortex2_bringup/launch/kortex_control.launch.py (hunk 1 above)
Comment (livanov93) [reply kortex2_bringup/launch/kortex_control.launch.py]: ab09b14

File: kortex2_controllers/src/fault_controller.cpp (hunk 5 above)
Comment (livanov93) [reply kortex2_controllers/src/fault_controller.cpp:117]: f301d3e

**Comment by livanov93** [comment]
@destogl This is ready to be merged, some ROS1 files and kinova API are not formatted well. This is already addressed in PR for removal of ROS1 stuff. Other parts will be fixed in additional PRs.

File: kortex2_controllers/CMakeLists.txt [hunk 6]
```diff
@@ -10,16 +10,18 @@ find_package(ament_cmake REQUIRED)
...
 set(THIS_PACKAGE_INCLUDE_DEPENDS
   controller_interface
   geometry_msgs
+  example_interfaces
 )
 
 include_directories(include)
 
-add_library(${PROJECT_NAME} SHARED src/twist_controller.cpp)
+add_library(${PROJECT_NAME} SHARED src/twist_controller.cpp src/fault_controller.cpp)
```
Comment (destogl) [inline kortex2_controllers/CMakeLists.txt:24]: I would add fault_controller to a separate library.

**Review by destogl — APPROVED** [review]
Looking good! Thanks for addressing my comments!

I would also propose to move fault controller to a separate library, but this is more like nitpick so you can do it here, in the follow-up or sometime in the future.

**Pull request merged**